Alternative docs (#52)
fox-it/dissect-docs: docs/source/rdump.rst

rdump
-----

> 

Reviewer: it's throwing off that the first line on rdump is not about rdump. 
maybe add an introduction line like the one in the tutorial
```suggestion
`rdump` is used to process, filter and format target-query results.
```

> By default target-query generates records. Records are binary representations of artefacts that

Reviewer 2: ```suggestion
By default ``target-query`` generates records. Records are binary representations of artefacts that
```

appear like this if they are transformed to text by the default mechanism:

.. code-block:: console

    <record key1="value1" key2="value2" >

> With rdump you can transform the stream of records to your liking. The rdump utility allows you to:

Reviewer 2: ```suggestion
With ``rdump`` you can transform the stream of records to your liking. The ``rdump`` utility allows you to:
```

* Select certain fields from the records.
* Filter certain records.
* Create additional derived fields.
* Limit the output.
* Format the results.
* Write the results through an adapter.

.. hint::

    Don't know yet what a record is? Read more :ref:`here <overview/index:records>` for a detailed explanation.

Field Selection
~~~~~~~~~~~~~~~

Imagine we're going to extract a list of users from a forensic image using:

.. code-block:: console

    $ target-query host.img -f users 

> To select only the name and home fields we use the ``-F`` option:

Reviewer: ```suggestion
To select only the 'name' and 'home' fields we use the ``-F`` option:
```

.. code-block:: console

    $ target-query host.img -f users | rdump -F name,home

Using the ``-X`` option we can exclude fields:

>     $ target-query host.img -f users | rdump -X hostname,domain,sid

Reviewer 2: I would love to see the output of all three commands. This will make it for the reader way easier to understand the implications of the Field selections.

Reviewer 2: Whilst reading further I would say this comment applies to this whole file

Author: No, I don't think that is necessary. It also creates too much visual clutter.

Filtering
~~~~~~~~~

Filtering records can be done through the ``-s`` option.

> The selection option must be Python-expression, where the record is represented by the symbol ``r``.

Reviewer 2: ```suggestion
The selection option must be a Python-expression, where the record is represented with the symbol ``r``.
```

> So to eliminate all records that have no domain value:

Reviewer 2: ```suggestion
So to eliminate all records that for example have no domain value:
```

> So to eliminate all records that for example have no domain value:

Reviewer: ```suggestion
So to eliminate all records that, for example, have no domain value:
```

> 

Reviewer: when you filter on a field you didn't select, will it show you that field value in the output? 
for example here, will it show the domain? or do you need to select it? maybe worth mentioning

Author: I presume this is common knowledge but I added a note, just to be sure.

.. code-block:: console

    $ target-query host.img -f users | rdump -F name,home -s "r.domain is not None"

Derivations
~~~~~~~~~~~

It is possible to create new fields derived from other fields. This can be done by using
an expression and the ``-E`` option. The following example lists all dll files from
a host and adds the filename to a separate field:

> 

Reviewer: do you have to make sure you don't use a name of an existing field? how do you know which fields are included in the record?

Author: It is quite obvious that you can use target-query for this. Also, yes, all fields must exist. This is common knowledge.

.. code-block:: console

    $ target-query host.img -f walkfs | rdump -s "r.path.suffix=='.dll'" -F path,file,size -E "file=path.name"
    
    <filesystem/entry path='\sysvol\Program Files\Ethereal\libgdk_pixbuf-2.0-0.dll' size=0.11 MB file='libgdk_pixbuf-2.0-0.dll'>
    <filesystem/entry path='\sysvol\Program Files\Ethereal\libgmodule-2.0-0.dll' size=26.6 KB file='libgmodule-2.0-0.dll'>
    <filesystem/entry path='\sysvol\Program Files\Ethereal\libgobject-2.0-0.dll' size=0.25 MB file='libgobject-2.0-0.dll'>
    ....and many more....

Limiting
~~~~~~~~

The example above yields quite a lot records. Limiting the number of records can be done through

> the ``--count`` and ``--skip`` options:

Reviewer: what does each of them do separately? is it common knowledge? I understand you count 3 lines. what does skip do? skip the rest? what is the meaning of the 1 value?

Author: This is common knowledge for most readers.

.. code-block:: console

    $ target-query host.img -f walkfs | rdump -s "r.path.suffix=='.dll'" -F path,file,size -E "file=path.name" --skip=1 --count=3

    <filesystem/entry path='\sysvol\My Documents\COMMANDS\cygwinb19.dll' size=0.38 MB file='cygwinb19.dll'>
    <filesystem/entry path='\sysvol\My Documents\ENUMERATION\NT\Cerberus\dnsscan.dll' size=40.0 KB file='dnsscan.dll'>
    <filesystem/entry path='\sysvol\My Documents\ENUMERATION\NT\Cerberus\fingerscan.dll' size=40.0 KB file='fingerscan.dll'>

Formatting
~~~~~~~~~~

A Python-style formatting rule can be specified using the ``-f`` option. The following example applies the format
``{file} ({size})`` to the records:

.. code-block:: console

>     $ target-query host.img -f walkfs | rdump -s "r.path.suffix=='.dll'" -F path,file,size -E "file=path.name" --skip=1 --count=3 -f "{file} ({size})"
>     
>     cygwinb19.dll (0.38 MB)
>     dnsscan.dll (40.0 KB)
>     fingerscan.dll (40.0 KB)

Reviewer 2: ```suggestion
    $ target-query host.img -f walkfs | rdump -s "r.path.suffix=='.dll'" -F path,file,size -E "file=path.name" --skip=1 --count=3 -f "Filename: {file} Size: {size}"
    
    Filename: cygwinb19.dll Size: 0.38 MB
    Filename: dnsscan.dll Size: 40.0 KB)
    Filename: fingerscan.dll Size: 40.0 KB
```
Would likely make the format and the customization possibilities more understandable.

> Instead of having to design your own format you can also choose one of these

Reviewer: ```suggestion
Instead of having to design your own format you can also choose one of these per-defined formats:
```

* JSON (``-j`` or ``--mode=json``)
* CSV (``-C`` or ``--mode=csv``)
* Line (``-L`` or ``--mode=line``)

> 

Reviewer: all the other bullets got their own section
```suggestion
Adapter Formatting
~~~~~~~~~~~~
```

It is also possible let an adapter take care of the formatting. For instance, if you wish to have your
records in an archive format with a year-month-day folder structure, you can employ the ``-w`` option and

> choose the archive adapter: ``archive://outputdir``. For complete list of adapters use ``-a``.

Reviewer 2: Great that the -w parameter is discussed, however I would propose to highlight another adapter that an analyst is more likely to use. Such as Splunk or Elastic.

Author: But also more complicated, as it requires additional knowledge about these systems. While this adapter is less frequently used, it is more suitable for demonstration purposes. The rest can be covered in the advanced sections.

> records in an archive format with a year-month-day folder structure, you can employ the ``-w`` option and

Reviewer: maybe add an example of the output of -w ? are adapters something we provide?
